plugins/modules/packaging/os/pacman.py

```diff
@@ -273,68 +288,96 @@ def run(self):
 
     def install_packages(self, pkgs):
         pkgs_to_install = []
+        pkgs_to_install_from_url = []
         for p in pkgs:
+            if p.source_is_URL:
+                pkgs_to_install_from_url.append(p)
+                continue
             if (
                 p.name not in self.inventory["installed_pkgs"]
                 or self.target_state == "latest"
                 and p.name in self.inventory["upgradable_pkgs"]
             ):
                 pkgs_to_install.append(p)
 
-        if len(pkgs_to_install) == 0:
+        if len(pkgs_to_install) == 0 and len(pkgs_to_install_from_url) == 0:
             self.add_exit_infos("package(s) already installed")
             return
 
-        self.changed = True
         cmd_base = [
             self.pacman_path,
-            "--sync",
             "--noconfirm",
             "--noprogressbar",
             "--needed",
         ]
         if self.m.params["extra_args"]:
             cmd_base.extend(self.m.params["extra_args"])
 
-        # Dry run first to gather what will be done
-        cmd = cmd_base + ["--print-format", "%n %v"] + [p.source for p in pkgs_to_install]
-        rc, stdout, stderr = self.m.run_command(cmd, check_rc=False)
-        if rc != 0:
-            self.fail("Failed to list package(s) to install", stdout=stdout, stderr=stderr)
+        def _build_install_diff(pacman_verb, pkglist):
+            # Dry run to build the installation diff
 
-        name_ver = [l.strip() for l in stdout.splitlines()]
+            cmd = cmd_base + [pacman_verb, "--print-format", "%n %v"] + [p.source for p in pkglist]
+            rc, stdout, stderr = self.m.run_command(cmd, check_rc=False)
+            if rc != 0:
+                self.fail("Failed to list package(s) to install", stdout=stdout, stderr=stderr)
+
+            name_ver = [l.strip() for l in stdout.splitlines()]
+            before = []
+            after = []
+            to_be_installed = []
+            for p in name_ver:
+                # With Pacman v6.0.1 - libalpm v13.0.1, --upgrade outputs "loading packages..." on stdout. strip that
+                if "loading packages" in p:
```

Comment:
This is annoying, but looking at pacman's source it looks like they use `printf` everywhere for "normal" messages.
`pacman --upgrade <pkg>` seems like the only path where this module hits this behavior however.

I really wished it output its information stuff on stderr, especially when using `--print-format` which, I assume, is designed to be used with scripts, but here we are, parsing the output to strip unwanted messages.

I think this case is safe since the module sets `LC_ALL=C` when starting, so even though the string is translated, it should only ever be in english when pacman is called from the module.

Reply:
You migth want to set `LANG='C', LC_ALL='C', LC_MESSAGES='C', LC_CTYPE='C'` instead of just `LC_ALL='C'`. This is what most modules do that expect non-translated messages.

Reply:
From what I could gather, `LANG` sets the default locale, that is, `LC_*` variables that are not set will take their value from `LANG`, whereas `LC_ALL` overrides all `LC_*` variables.

In this specific case, I think it is safe to only set `LC_ALL` since `pacman` calls [`setlocale(LC_ALL, "")`](https://gitlab.archlinux.org/pacman/pacman/-/blob/a8d9fb8d9c57d96cc92373560f1d32b25a97c160/src/pacman/pacman.c#L256-265), which, according to the [manpage](https://man.archlinux.org/man/setlocale.3), will set all `LC_*` variables to the value of `LC_ALL`.

```
jean@x1:~$ LANG=fr_FR.utf8 pacman --upgrade --noconfirm --noprogressbar --needed --print-format "%n %v" /var/cache/pacman/pkg/zsnes-2.0.9-1-x86_64.pkg.tar.zst /var/cache/pacman/pkg/pipewire-jack-1:0.3.45-1-x86_64.pkg.tar.zst  2>/dev/null
chargement des paquets…
zsnes 2.0.9-1
pipewire-jack 1:0.3.45-1
jean@x1:~$ LANG=fr_FR.utf8 LC_ALL=C pacman --upgrade --noconfirm --noprogressbar --needed --print-format "%n %v" /var/cache/pacman/pkg/zsnes-2.0.9-1-x86_64.pkg.tar.zst /var/cache/pacman/pkg/pipewire-jack-1:0.3.45-1-x86_64.pkg.tar.zst  2>/dev/null
loading packages...
zsnes 2.0.9-1
pipewire-jack 1:0.3.45-1
```

- https://sourceware.org/glibc/wiki/Locales#Using_Locales
- https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html#tag_08_02
- https://man.archlinux.org/man/setlocale.3